shinysdr/plugins/wspr/demodulator.py

```diff
@@ -206,9 +218,11 @@ def fileClosed(self, filename):
 
         # wsprd expects its -f argument to be as if the recording was made with
         # a USB receiver
-        dial_freq = (self.__start_frequency - self.audio_frequency) / 1e6
+        dial_freq = (rf_frequency - self.audio_frequency) / 1e6
+        self.__decoder_active = defer.Deferred()
+        self.__decoder_active.addCallbacks(self.__decode_finished, self.__decode_finished)
```

Comment:
I think this is a silly antipattern. Deferreds are not event handlers: they are a thing you return from a function representing the result that isn't available now. Hence the name.

Consider this:

```python
def do_something(on_success, on_error):
    try:
        result = something()
    except Exception as e:
        on_error(e)
    else:
        on_success(result)
```

Ewww, callbacks! Let's make that:

```python
def do_something(d):
    try:
        result = something()
    except Exception as e:
        d.errback(e)
    else:
        d.callback(result)
```

That's much better, right? But then why not simply use the first implementation with callbacks, and do this:

```
d = Deferred()
do_something(d.callback, d.errback)
```

How exactly is this better? If you were going to have a deferred here, it would be something returned from a function like `get_wsprd_result()`. But there's no need, because there's already an implementation of `IProcessProtocol` here which has given names to all the kinds of events that could happen from a process.

The pattern here isn't materially different than the common case of a TCP server with a factory and a protocol. The protocol gets events that happen on the connection. The protocol has a reference to the factory as `self.factory`. When something happens on a connection that's relevant to the global state, the protocol calls `self.factory.something_happened(...)`. There are usually no deferreds.

Consider, in all of the Twisted code, excluding tests, documentation, and defer.py itself, there are only 103 places where a deferred is created.

```console
$ grep -I --exclude 'defer.py' --exclude '*test*' -r 'Deferred()' src/twisted/ | wc -l
103
```

If deferreds were a token for "let me know when this thing happens", that number would be _much_ higher. Look through those 103 instances, and it's almost always a function creating a deferred and then returning it. Occasionally the deferred gets passed between methods of a class through a private member.